instrumentation/netty/netty-4-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/common/client/AbstractNettyHttpClientTracer.java

```diff
@@ -84,39 +89,58 @@ protected String responseHeader(HttpResponse httpResponse, String name) {
     return SETTER;
   }
 
-  public static void operationComplete(Future<?> future) {
-    AbstractNettyHttpClientTracer tracer = NettyHttpClientTracerAccess.getTracer();
-    if (tracer == null) {
-      return;
+  public Context startConnectionSpan(Context parentContext, SocketAddress remoteAddress) {
+    if (!alwaysCreateConnectSpan) {
+      return null;
     }
 
-    if (!(future instanceof ChannelFuture)) {
-      return;
-    }
-    // If first call to GenericFutureListener#operationComplete has an exception then we
-    // treat it as the cause of connection failure and create a special span for it
-    ChannelFuture channelFuture = (ChannelFuture) future;
-    Context parentContext = tracer.getAndRemoveConnectContext(channelFuture);
-    if (parentContext == null) {
-      return;
-    }
-    Throwable cause = future.cause();
-    if (cause == null) {
-      return;
-    }
+    SpanBuilder spanBuilder = spanBuilder(parentContext, "CONNECT", INTERNAL);
+    NetPeerAttributes.INSTANCE.setNetPeer(spanBuilder, (InetSocketAddress) remoteAddress);
+
+    return parentContext.with(spanBuilder.startSpan());
+  }
 
-    if (tracer.shouldStartSpan(parentContext, SpanKind.CLIENT)) {
-      tracer.connectionFailure(parentContext, channelFuture.channel(), cause);
+  public void endConnectionSpan(
+      Context context,
+      Context parentContext,
+      SocketAddress remoteAddress,
+      Channel channel,
+      Throwable throwable) {
+    if (context != null) {
```

Comment:
Nit: using `alwaysCreateConnectSpan` will probably allow JIT to skip the `else` part easily when that property is true

Reply:
Added `alwaysCreateConnectSpan` check, though I suspect that it won't change much for jit as I believe it speculatively removes branches that have not executed anyway.